File: agent/commands-generator/src/main/java/org/matilda/commands/CommandsGeneratingAnnotationProcessor.java
package org.matilda.commands;

import org.matilda.commands.di.AnnotationProcessorModule;
import org.matilda.commands.di.CommandsGeneratorComponent;
import org.matilda.commands.di.DaggerCommandsGeneratorComponent;
import org.matilda.commands.exceptions.AnnotationProcessingException;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;
import java.util.Set;

import static org.matilda.commands.python.PythonProperties.PYTHON_GENERATED_PACKAGE_OPTION;
import static org.matilda.commands.python.PythonProperties.PYTHON_ROOT_DIR_OPTION;

@SupportedSourceVersion(SourceVersion.RELEASE_8)
@SupportedAnnotationTypes({"org.matilda.commands.MatildaService", "org.matilda.commands.MatildaCommand"})
@SupportedOptions({PYTHON_ROOT_DIR_OPTION, PYTHON_GENERATED_PACKAGE_OPTION})
public class CommandsGeneratingAnnotationProcessor extends AbstractProcessor {
    private ProcessingEnvironment mProcessingEnvironment;
    private boolean mWasRun = false;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        mProcessingEnvironment = processingEnv;
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        try {
            CommandsGeneratorComponent component = DaggerCommandsGeneratorComponent.builder()
                    .annotationProcessorModule(new AnnotationProcessorModule(annotations, roundEnv,
                            mProcessingEnvironment, mWasRun))
                    .build();
            component.commandsGenerator().generate();
            mWasRun = true;
        } catch (AnnotationProcessingException e) {
            e.printStackTrace();
            mProcessingEnvironment.getMessager().printMessage(Diagnostic.Kind.ERROR, e.getMessage(), e.element);
        } catch (Throwable e) {
            e.printStackTrace();
            mProcessingEnvironment.getMessager().printMessage(Diagnostic.Kind.ERROR, e.getMessage());
        }
        return false;
    }
}

File: agent/commands-generator/src/main/java/org/matilda/commands/CommandsGeneratingAnnotationProcessor.kt
package org.matilda.commands

import androidx.room.compiler.processing.*
import androidx.room.compiler.processing.javac.JavacBasicAnnotationProcessor
import org.matilda.commands.di.AnnotationProcessorModule
import org.matilda.commands.di.DaggerCommandsGeneratorComponent
import org.matilda.commands.exceptions.AnnotationProcessingException
import org.matilda.commands.java.JavaProperties
import org.matilda.commands.protobuf.ProtobufLocations
import org.matilda.commands.python.PythonProperties
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.tools.Diagnostic

@ExperimentalProcessingApi
@SupportedSourceVersion(SourceVersion.RELEASE_8)
@SupportedAnnotationTypes("org.matilda.commands.MatildaService", "org.matilda.commands.MatildaCommand",
    "org.matilda.commands.MatildaDynamicService")
@SupportedOptions(
    PythonProperties.PYTHON_ROOT_DIR_OPTION,
    PythonProperties.PYTHON_GENERATED_PACKAGE_OPTION,
    ProtobufLocations.PROTOBUF_DIRS_OPTION,
    JavaProperties.JAVA_MAIN_PACKAGE_OPTION,
)
class CommandsGeneratingAnnotationProcessor : JavacBasicAnnotationProcessor(), XProcessingStep {
    private var mProcessingEnvironment: XProcessingEnv? = null
    private var mWasRun = false
    @Synchronized
    override fun initialize(env: XProcessingEnv) {
        mProcessingEnvironment = env
    }

    override fun processingSteps(): Iterable<XProcessingStep> {
        return listOf(this)
    }

    override fun annotations(): Set<String> = setOf("org.matilda.commands.MatildaService",
        "org.matilda.commands.MatildaCommand",
        "org.matilda.commands.MatildaDynamicService")

    override fun preRound(env: XProcessingEnv, round: XRoundEnv) {
        try {
            val component = DaggerCommandsGeneratorComponent.builder()
                .annotationProcessorModule(AnnotationProcessorModule(env, round, mWasRun))
                .build()
            component.commandsGenerator().generate()
            mWasRun = true
        } catch (e: AnnotationProcessingException) {
            e.printStackTrace()
            env.messager.printMessage(Diagnostic.Kind.ERROR, e.message ?: "", e.element)
        } catch (e: Throwable) {
            e.printStackTrace()
            env.messager.printMessage(Diagnostic.Kind.ERROR, e.message ?: "")
        }
    }
}
